### PR TITLE
Spectrogram front end for Keras

### PR DESCRIPTION
**Pull request by matt-har-vey**
The module added in this change wraps the transformation of a raw
waveform to an image input for a convolutional neural network in a Keras
layer. This can enable training end-to-end from the raw waveform to
class probabilities or, alternatively, serve as a module to reliably
apply the same spectrogram settings in preprocessing and inference.

Change-Id: Ic9b51d3c288baf1faddb38ae95762555352dafe9

### CODE REVIEW

File: multispecies_whale_detection/front_end.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+
+def amplitude_ratio_to_db(x):
+  return 20.0 * tf.math.log(x / 10.0)
+
+
+def db_to_amplitude_ratio(x):
+  return tf.math.pow(10.0, x / 20.0)
+
+
+@dataclasses.dataclass(frozen=True)
```
Comment (sdenton4) [inline multispecies_whale_detection/front_end.py:31]: I've always started with frozen=True and ended up dropping it to either enable easier debugging and/or parameter sweeps (eg, start from a 'base' and then update a particular param.)

File: multispecies_whale_detection/front_end.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+class FrontEndConfig:
+  """Configuration data object for the FrontEnd Keras layer."""
+
+  sample_rate: int = 2000
+  """Sample rate of the input waveform."""
```
Comment (sdenton4) [inline multispecies_whale_detection/front_end.py:61]: piconit: Why not a single-line comment? (For some reason putting the comment above the variable also seems more intuitive to me, but no hard preference.)

File: multispecies_whale_detection/front_end.py [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+  frame_length = int(frame_seconds * sample_rate)
+  frame_step = int(hop_seconds * sample_rate)
+  tf.assert_greater(frame_length, 0)
+  tf.assert_greater(frame_step, 0)
+  stft = tf.signal.stft(
+      waveform, frame_length=frame_length, frame_step=frame_step, pad_end=False)
+  return tf.abs(stft)
+
+
+class FrontEnd(tf.keras.layers.Layer):
```
Comment (sdenton4) [inline multispecies_whale_detection/front_end.py]: Consider something more specific, like LogSpectrogramFrontEnd

File: multispecies_whale_detection/front_end.py [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+  return 20.0 * tf.math.log(x / 10.0)
+
+
+def db_to_amplitude_ratio(x):
+  return tf.math.pow(10.0, x / 20.0)
+
+
+@dataclasses.dataclass(frozen=True)
+class FrontEndConfig:
+  """Configuration data object for the FrontEnd Keras layer."""
```
Comment (sdenton4) [inline multispecies_whale_detection/front_end.py]: Another alternative is to add documentation in the dataclass docstring, instead of comments beside individual params. This matches style for method/function docs, at least.

File: multispecies_whale_detection/front_end.py [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+  It optionally normalizes by subtracting a per-channel (frequency bin) noise
+  floor, which is estimated as a percentile of the STFT magnitudes in that
+  channel.
+  """
+
+  def __init__(self, config=None):
+    """Initializes this layer.
+
+    Args:
+      config:
```
Comment (sdenton4) [inline multispecies_whale_detection/front_end.py]: Looks like this doc string was left hanging.

File: multispecies_whale_detection/front_end.py (hunk 1 above)
Comment (matt-har-vey) [reply multispecies_whale_detection/front_end.py:31]: I think starting from frozen=True is good practice. Heading off spooky action at a distance is worth a lot.

Does "debugging" mean ease of not-to-be-committed changes during development?

dataclasses.replace is satisfying to me for hyperparameter sweeps, and it's only a little bit of extra typing for the debugging use case.

File: tests/test_front_end.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+import tensorflow_probability as tfp
+
+
+class TestFrontEnd(unittest.TestCase):
+
+  def test_call_on_noise(self):
+    sample_rate = 2000
+    config = front_end.FrontEndConfig(sample_rate=sample_rate,)
+    num_samples = sample_rate * 5
+    noise_waveform = tf.random.normal([num_samples], 0, 1e-3)
```
Comment (sdenton4) [inline tests/test_front_end.py:32]: May want to set a seed here, as well.

File: multispecies_whale_detection/front_end.py [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+        waveform,
+        sample_rate=sample_rate,
+        frame_seconds=self._config.frame_seconds,
+        hop_seconds=self._config.hop_seconds)
+    num_frequency_bins = magnitude.shape[-1]
+    hz_per_bin = sample_rate / 2 / num_frequency_bins
+
+    db = amplitude_ratio_to_db(magnitude + self._config.log_stabilizer)
+
+    if self._config.per_channel_normalization:
```
Comment (sdenton4) [inline multispecies_whale_detection/front_end.py]: It's easy to imagine this boolean becoming an enum (eg, if you want to try PCEN); may be worth getting ahead of it.

Also may be worth breaking the smoothing code out into its own method. When signal processing logic stacks up in the 'call' method it can get hard to read/reason about.

File: multispecies_whale_detection/front_end.py (hunk 4 above)
Comment (matt-har-vey) [reply multispecies_whale_detection/front_end.py]: Let's roll this one into the other thread on the same theme.

File: multispecies_whale_detection/front_end.py (hunk 2 above)
Comment (matt-har-vey) [reply multispecies_whale_detection/front_end.py:61]: Some doc generators recognize this syntax as an attribute docstring (hearsay about Sphinx and [PyCharm](https://youtrack.jetbrains.com/issue/PY-28549#focus=Comments-27-4299152.0-0)). I don't find the ordering intuitive either, but it is at least consistent with the way it's done for classes and functions, and there are some semi-official positions statements that it's conventional.

"""
String literals occurring immediately after a simple assignment at the top level of a module, class, or init method are called "attribute docstrings".
"""

https://www.python.org/dev/peps/pep-0257/

File: multispecies_whale_detection/front_end.py (hunk 3 above)
Comment (matt-har-vey) [reply multispecies_whale_detection/front_end.py]: Partially done.

I definitely agree that something with _Spectrogram_ is better (more descriptive, doesn't stutter, not necessarily a front end).

I'd rather not be as specific as _Log_ because I anticipate configurable compression type to come soon (e.g. power-law independent of PCEN, or PCEN itself, where stabilized power-law compression is applied after the normalization).

File: multispecies_whale_detection/front_end.py (hunk 5 above)
Comment (matt-har-vey) [reply multispecies_whale_detection/front_end.py]: Done. Thanks.

File: tests/test_front_end.py (hunk 6 above)
Comment (matt-har-vey) [reply tests/test_front_end.py:32]: Done.

I don't feel strongly about it, but in cases where the assertions are will hold for any choice of seed, I intentionally omit setting one, since it makes the assertions stronger. I recognize the trade off that future modifications may assume a seed had been sit, not have detailed enough familiarity to realize it's necessary, etc.

**Review by sdenton4 — APPROVED** [review]
(Suggestions are non-blocking; feel free to integrate per your judgement.)

File: multispecies_whale_detection/front_end.py (hunk 7 above)
Comment (matt-har-vey) [reply multispecies_whale_detection/front_end.py]: Done.

There's still some ugly manual bits in the config serialization code, and I'm getting tempted to just adpot Keras more completely and make Layer the base class and serialization object for normalization strategies and then pass a normalization Layer as an optional kwarg to Spectrogram.__init__.

That seems to head toward recreating more of [LEAF](https://github.com/google-research/leaf-audio/tree/master/leaf_audio), though. I don't want to go too far with this implementation if it is going to end up being just a placeholder.

While I don't think "learnability" will be critical, there are some reusable spectrogram and PCEN layers in there. A few things I'd like that it doesn't have are: be in pypi, have tests, have an option for linear frequency scaling.
